examples/cookbook/recipe_device_allocator.cpp

```diff
@@ -0,0 +1,98 @@
+//////////////////////////////////////////////////////////////////////////////
+// Copyright (c) 2016-21, Lawrence Livermore National Security, LLC and Umpire
+// project contributors. See the COPYRIGHT file for details.
+//
+// SPDX-License-Identifier: (MIT)
+//////////////////////////////////////////////////////////////////////////////
+#include <stdio.h>
+
+#include "umpire/util/device_allocator_helper.hpp"
+#include "umpire/ResourceManager.hpp"
+
+const int THREADS_PER_BLOCK = 128;
+const int BLOCKS = 4;
+__device__ int* data{nullptr};
+
+/*
+ * Kernel that performs simple arithmetic and stores the
+ * result in the array of ints passed to the kernel.
+ *
+ * Note: Benign data race that writes to the device allocated
+ * array, data.
+ */
+__global__ void my_kernel(int* result)
+{
+  unsigned int i{threadIdx.x+blockIdx.x*blockDim.x};
+
+  if (i == 0) {
+    umpire::DeviceAllocator alloc = umpire::util::getDeviceAllocator(0);
+    data = static_cast<int*>(alloc.allocate(BLOCKS * sizeof(int)));
+  }
+
+  data[blockIdx.x] = blockIdx.x;
```

Comment:
What happens when threads with non-zero ids (`i`) write to `data` before data is allocated by the thread with id of zero?